File: clients/OpenSync.Client.Js/src/SyncChannel.ts
import { OpenSyncClient, SyncEvent } from './OpenSyncClient';

export class SyncChannel {
  private client: OpenSyncClient;
  private name: string;

  constructor(client: OpenSyncClient, name: string) {
    this.client = client;
    this.name = name;
  }

  async subscribe(): Promise<void> {
    await this.client.subscribe('channel', this.name);
  }

  async join(metadata?: any): Promise<void> {
    await this.client.httpPost(`/channels/${this.name}/members`, { identity: 'default', metadata });
  }

  async leave(): Promise<void> {
    await this.client.httpPost(`/channels/${this.name}/leave`, {});
  }

  async sendMessage(data: any): Promise<void> {
    await this.client.httpPost(`/channels/${this.name}/messages`, { data });
  }

  onMemberJoined(handler: (data: any) => void): void {
    this.client.on('member_joined', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }

  onMemberLeft(handler: (data: any) => void): void {
    this.client.on('member_left', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }

  onMessage(handler: (data: any) => void): void {
    this.client.on('message_broadcast', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }
}

File: nupkgs/OpenSync.Sdk/SyncChannel.cs
using OpenSync.Sdk.Events;

namespace OpenSync.Sdk;

public class SyncChannel
{
    private readonly OpenSyncClient _client;
    private readonly string _uniqueName;

    public event EventHandler<PresenceChangedEventArgs>? OnMemberJoined;
    public event EventHandler<PresenceChangedEventArgs>? OnMemberLeft;
    public event EventHandler<PresenceChangedEventArgs>? OnPresenceUpdated;
    public event EventHandler<MessageEventArgs>? OnMessage;

    public SyncChannel(OpenSyncClient client, string uniqueName)
    {
        _client = client;
        _uniqueName = uniqueName;
        _client.OnEvent += HandleEvent;
    }

    private void HandleEvent(object? sender, SyncEventArgs args)
    {
        if (args.ObjectType != "channel" || args.ObjectId != _uniqueName) return;

        switch (args.Event)
        {
            case "member_joined": OnMemberJoined?.Invoke(this, new PresenceChangedEventArgs(args.Data)); break;
            case "member_left": OnMemberLeft?.Invoke(this, new PresenceChangedEventArgs(args.Data)); break;
            case "presence_updated": OnPresenceUpdated?.Invoke(this, new PresenceChangedEventArgs(args.Data)); break;
            case "message_broadcast": OnMessage?.Invoke(this, new MessageEventArgs(args.Data)); break;
        }
    }

    public async Task JoinAsync(object? metadata = null)
    {
        var json = metadata != null ? System.Text.Json.JsonSerializer.Serialize(metadata) : null;
        await _client.HttpPostAsync<object>($"/channels/{_uniqueName}/members", new { identity = "default", metadata = json });
    }

    public async Task LeaveAsync()
        => await _client.HttpPostAsync<object>($"/channels/{_uniqueName}/leave", new { });

    public async Task SendMessageAsync(object data)
    {
        var json = System.Text.Json.JsonSerializer.Serialize(data);
        await _client.HttpPostAsync<object>($"/channels/{_uniqueName}/messages", new { data = json });
    }

    public async Task SubscribeAsync() => await _client.SubscribeAsync("channel", _uniqueName);
}
